### PR TITLE
[SecurityBundle] Allow for custom request matchers

### PR DESCRIPTION
**Pull request by ro0NL**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes/no |
| Fixed tickets | #20257 |
| License | MIT |
| Doc PR | reference to the documentation PR, if any |

Quick proof of concept to allow for custom request matchers in the security bundle.

This would allow for matching a request based on a route name (routing happens before security!).

With only this setting it would still be cumbersome to do it though;

``` yml
access_control:
  - { matcher: service_matching_route1 }
  - { matcher: service_matching_route2 }
  # etc
```

To solve this we could allow arguments or have it built in:

``` yml
access_control:
  - { matcher: generic_route_matcher, matcher_arguments: [route1] }
  - { route: route2 }
```

edit: added `ChainRequestMatcher` as it could be useful to have a complex matcher built from simple matchers. Allows to refactor the current `RequestMatcher` to be built from `MethodRequestMatcher`, `SchemeRequestMatchter`, etc. I would like this kind of reuseability a lot :)

edit2: added separated PR for it (#20274)

### CODE REVIEW

**Comment by ro0NL** [comment]
To clarify; the real issue is i want to target a specific controller/action, without depending on `@Security()` annotation.

Currently we need to duplicate a path definition as a regex ourselves (instead of relying on the compiled regex out-of-the-box) which is error prone.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php [hunk 1]
```diff
@@ -163,7 +163,9 @@ private function addAccessControlSection(ArrayNodeDefinition $rootNode)
                     ->prototype('array')
                         ->fixXmlConfig('ip')
                         ->fixXmlConfig('method')
+                        ->fixXmlConfig('role')
                         ->children()
+                            ->scalarNode('matcher')->defaultNUll()->end()
```
Comment (iltar) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: `defaultNUll` should be `defaultNull` 😄

File: src/Symfony/Component/HttpFoundation/ChainRequestMatcher.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpFoundation;
+
+/**
+ * @author Roland Franssen <franssen.roland@gmail.com>
+ */
+class ChainRequestMatcher implements RequestMatcherInterface
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/ChainRequestMatcher.php]: Is this class related to this PR? It's not used anywhere?
About the name, "chain" usually means "delegation chain", ie the first applicable in the chain takes responsibility for the job. Here, this is something else, like `AllMatchersRequestMatcher` (dummy suggestion, I'm not really proud of it :) )

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php [hunk 3]
```diff
@@ -180,9 +182,6 @@ private function addAccessControlSection(ArrayNodeDefinition $rootNode)
                                 ->prototype('scalar')->end()
                             ->end()
                             ->scalarNode('allow_if')->defaultNull()->end()
-                        ->end()
-                        ->fixXmlConfig('role')
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: this CS should be applied to a lower branch

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 4]
```diff
@@ -195,13 +195,11 @@ private function createAuthorization($config, ContainerBuilder $container)
...
         foreach ($config['access_control'] as $access) {
-            $matcher = $this->createRequestMatcher(
-                $container,
-                $access['path'],
-                $access['host'],
-                $access['methods'],
-                $access['ips']
-            );
+            if (null !== $access['matcher']) {
+                $matcher = new Reference($access['matcher']);
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: the currently created matcher is configured by many props in `$access`. Thus, is this extension point really enough when you don't give it these parameters?

File: src/Symfony/Component/HttpFoundation/ChainRequestMatcher.php (hunk 2 above)
Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/ChainRequestMatcher.php]: Nope. It's outdate-ish. I moved it to #20274.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php (hunk 4 above)
Comment (ro0NL) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: Im not sure about the "right" approach here. Perhaps we should talk config/API first.

But we could do `matcher|<combination of other keys>`.

DX-wise i think `- { route: foo, roles: [] }` is best and probably same for firewalls? (using an array of routes).

As goes for `- { expression: foo }`. But that could work counter-intuitive with `allow_if`.

**Comment by ro0NL** [comment]
Closing in favor of #20274, im not sure about the matcher approach here. Maybe something better comes out. The integration is definitely missing though.

But, im going to experiment with `path: '%route.name%'` or so, keeping config as is, while still allowing me to reference a path regex (compiled i guess) by name. It would be the cleanest solution i think.

**Pull request closed without merging**